src/lib/lossless_cast.hpp

```diff
@@ -0,0 +1,229 @@
+#pragma once
+
+#include <optional>
+#include <type_traits>
+
+#include "all_type_variant.hpp"
+#include "null_value.hpp"
+
+/**
+ * This file defines the following casting functions that guarantee **lossless** conversion between data types and
+ * return std::nullopt if no lossless conversion was possible.
+ *      lossless_cast<Target>(Source)
+ *      lossless_variant_cast<Target>(AllTypeVariant)
+ *      lossless_variant_cast(Source, DataType)
+ *
+ * "Lossless" means just that: No information must be lost during casting. Typical lossful operations would be
+ *      - float-to-int casts with the float having a non-zero fractional part
+ *      - long-to-int casts where the long was > MAX_INT
+ *      - string-to-int where the string contained no number, or a number > MAX_INT.
+ *      - double-to-float where the significand of the double has bits set that cannot be represented in float
+ *
+ * Whenever possible you should use these casts over "comfortable" casts such as static_cast or boost::lexical_cast.
+ * Those might incur information loss and might thus introduce bugs. See #1306, #1525.
+ *
+ * THINK ABOUT AN ALGORITHM'S DATA, THE POSSIBLE STATES IT MIGHT HAVE, AND HANDLE THEM PROPERLY.
+ */
+
+namespace opossum {
+
+inline std::tuple<bool, int32_t, uint32_t> decompose_floating_point(float f) {
+  auto* bits = reinterpret_cast<uint32_t*>(&f);
+
+  auto exponent = static_cast<int32_t>(((*bits) >> 23) & 0xFF);
+  auto fraction = (*bits) & 0x7FFFFF;
+  auto sign = ((*bits) & 0x8000) != 0;
+
+  return {sign, exponent, fraction};
+}
+
+inline std::tuple<bool, int32_t, uint64_t> decompose_floating_point(double f) {
+  auto* bits = reinterpret_cast<uint64_t*>(&f);
+
+  auto exponent = static_cast<int32_t>(((*bits) >> 52) & 0x7FF);
+  auto fraction = (*bits) & 0xFFFFFFFFFFFFF;
+  auto sign = ((*bits) & 0x80000000) != 0;
+
+  return {sign, exponent, fraction};
+}
+
+// Identity
+template <typename Target, typename Source>
+std::enable_if_t<std::is_same_v<Target, Source>, std::optional<Target>> lossless_cast(const Source& source) {
+  return source;
```

Comment:
Could universal references be beneficial here?

Reply:
idk, to what benefit? This is in a header, I'd expect this to get inlined anyway?

Reply:
Avoiding a string copy. Just a quick thought though.